src/fides/api/alembic/migrations/versions/5a8cee9c014c_privacy_preferences_v2_data.py

```diff
@@ -0,0 +1,377 @@
+    WHERE privacy_notice_history_id IS NOT NULL
+    ORDER BY created_at asc;
+"""
+
+
+class CurrentMigrationType(Enum):
+    preferences = "preferences"
+    served = "served"
+
+
+encryptor = sqlalchemy_utils.types.encrypted.encrypted_type.StringEncryptedType(
+    String,
+    CONFIG.security.app_encryption_key,
+    AesGcmEngine,
+    "pkcs5",
+)
+
+decryptor = sqlalchemy_utils.types.encrypted.encrypted_type.StringEncryptedType(
+    JSONTypeOverride,
+    CONFIG.security.app_encryption_key,
+    AesGcmEngine,
+    "pkcs5",
+)
+
+
+def migrate_current_records(
+    bind: Connection, starting_query: str, migration_type: CurrentMigrationType
+):
+    """Common method to migrate CurrentPrivacyPreference -> CurrentPrivacyPreferenceV2 and
+    LastServedNotice -> LastServedNoticeV2.
+
+    We are migrating from tables with unique constraints on two provided identity types x preferences types to the new
+    tables with unique constraints on email, device id, and phone number.
+
+    Migration involves linking all records in the original table with any shared identifiers across email, phone,
+    or device id, and collapsing these records into single rows, retaining the most recently used non-null identifiers
+    and recently saved preferences.
+    """
+    df: DataFrame = pd.read_sql(starting_query, bind)
+
+    if len(df.index) == 0:
+        logger.info(f"No {migration_type.value} records to migrate. Skipping.")
+        return
+
+    # Drop invalid rows where we have an encrypted val but not a hashed val and vice versa.
+    # Also drop if there are no identifiers at all.
+    # This would be unexpected, but this would mean our ProvidedIdentity record was not populated correctly.
+    df["email_count"] = df[["encrypted_email", "hashed_email"]].count(axis=1)
+    df["phone_count"] = df[["encrypted_phone", "hashed_phone_number"]].count(axis=1)
+    df["device_count"] = df[["encrypted_device", "hashed_fides_user_device"]].count(
+        axis=1
+    )
+    df = df[df["email_count"] != 1]
+    df = df[df["phone_count"] != 1]
+    df = df[df["device_count"] != 1]
+    df = df[df["email_count"] + df["phone_count"] + df["device_count"] >= 2]
+
+    # Create a "paths" column in the dataframe that is a list of non-null identifiers, so
+    # we only consider actual values as a match.
+    df["paths"] = df[
+        ["hashed_email", "hashed_phone_number", "hashed_fides_user_device"]
+    ].apply(lambda row: [val for val in row if pd.notna(val)], axis=1)
+
+    network_x_graph: nx.Graph = nx.Graph()
+    # Adds every path to the Graph
+    df["paths"].apply(lambda path: nx.add_path(network_x_graph, path))
+
+    # This is the magic - linking any common records across hashed_email OR hashed_phone OR hashed_device
+    connected_records: List[Set] = list(nx.connected_components(network_x_graph))
```

Comment:
This was the best I could come up with (and unfortunately installed another requirement to do it) and ended up doing a lot of computations outside of SQL with pandas.

The main thing, I needed to assign a `group_id` on records that had the same email OR phone OR device id and then collapse preferences on that group_id.

- If Record A has phone 1234 and device id apple
- Record B has device id apple
- Record C has email customer-1@example.com and device id apple, 
- Record D has phone 1234

This treats all of the above as the same person

```python
G = nx.Graph() 
nx.add_path(G, ["1234", "apple"])
nx.add_path(G, ["apple"])
nx.add_path(G, ["apple", "customer-1@example.com"])
nx.add_path(G, ["1234"])
nx.add_path(G, ["123", "customer"])
nx.add_path(G, ["customer"])
nx.add_path(G, ["orange"])

connected_records = list(nx.connected_components(G))
connected_records
```
```bash
[{'1234', 'apple', 'customer-1@example.com'}, {'123', 'customer'}, {'orange'}]
```

Reply:
thank you for spelling this out here, this is a very clever solution! it's bringing me back to a past life where i spent my days working with graph databases.

i have the same qualms about a new dependency/requirement for "just" a migration, but it does seem like the default dependences for networkx are _relatively light_ (i'm looking [here](https://github.com/networkx/networkx/blob/main/pyproject.toml#L57-L62), i assume that's the right place? it may be good to just confirm these don't bring in a bunch of new transitive dependencies we don't already have...) and this is an elegant way to meet a complex requirement - so i think it's worth it!

i wonder if it'd make sense to just make a note on the dependency itself in `requirements.txt` to indicate that it was only added for a migration - maybe we'd consider removing the migration and the dependency at some point in the future once we're confident everyone's been properly migrated away from this state and if/when we're trying to clean up dependencies for whatever reason.  (i know updating migration history is always a bit dicey, but it feels like it could be acceptable in this case?).